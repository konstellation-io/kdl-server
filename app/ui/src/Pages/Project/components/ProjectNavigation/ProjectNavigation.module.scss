@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';
@import 'Styles/shadows';

.container {
  @include shadow(4);
  width: 25 * $grid-unit;
  max-width: 5 * $grid-unit;
  background-color: palette(lowlight, 900);
  flex-shrink: 0;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  overflow: hidden;
  transition: max-width ease 0.3s;

  .bottom {
    background-color: palette(base);

    .navButton:hover {
      background-color: palette(base, 900);
    }
  }

  &.opened {
    max-width: 25 * $grid-unit;
  }
}
