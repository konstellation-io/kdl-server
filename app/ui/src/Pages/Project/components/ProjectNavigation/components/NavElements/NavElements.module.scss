@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.projectTools {
  background-color: palette(base, 300);
  a:active > div,
  a:hover > div,
  a:focus > div {
    background-color: palette(base, 200) !important;
  }

  .userTools {
    border-top: 2px solid palette(base, 400);
    margin-top: $grid-unit;
    transition: border-right 0.3s ease;
    border-right: 2px solid transparent;

    &.started {
      border-right: 2px solid palette(success);
    }
    &.stopped {
      border-right: 2px solid palette(alert);
    }
    &.loading {
      border-right: 2px solid palette(feedback);
    }

    .userToolLabel {
      @include font-caption;
      display: flex;
      color: font-color(regular);
      width: 100%;
      white-space: nowrap;
      margin-left: 2 * $grid-unit;
      @include circle-before(#3d3d3d);
      &.pending:before {
        background-color: palette(alert);
        top: $grid-unit - 5px;
      }
      &.running:before {
        background-color: palette(success);
        top: $grid-unit - 5px;
      }
      &.error:before {
        background-color: palette(error);
        top: $grid-unit - 5px;
      }
      padding-left: 3 * $grid-unit;
      flex-direction: row;
      justify-content: flex-start;
      align-items: center;
    }

    .usertoolsIcon {
      margin: $grid-unit;
      cursor: pointer;
      &.blocked {
        color: palette(alert);
      }
      &:active,
      &:hover,
      &:focus {
        color: palette(highlight);
      }
    }

    .usertoolsOptions {
      display: flex;
      align-items: center;
      justify-content: space-between;

      div[data-for="settings"],
      div[data-for="start"],
      div[data-for="stop"] {
        display: flex;
        justify-content: center;
        svg {
          width:  $grid-unit * 2;
          height: $grid-unit * 2;
        }
      }
      &.opened {
        height: 40px;
        div[data-for="settings"],
        div[data-for="start"],
        div[data-for="stop"] {
          svg {
            width: $grid-unit + 4;
            height: $grid-unit + 4;
          }
        }
      }
    }

    .usertoolsSettings {
      display: flex;
      flex-direction: column;
      width: 100%;
      .progressSpinnerContainer {
        margin-top: 0;
        margin-left: 13px;
      }
      &.opened {
        flex-direction: row;
        width: auto;
        .progressSpinnerContainer {
          margin-top: $grid-unit;
          margin-left: $grid-unit;
          margin-right: $grid-unit;
        }
      }
    }

    .toggleToolsWrapper {
      border-top: 2px dotted palette(base, 400);
      transition: opacity 0.3s ease;
      .loadingTools {
        color: palette(highlight);
        margin: 0 1.4 * $grid-unit;
        transition: transform ease 0.3s;
        flex-shrink: 0;
      }
      &.disabled {
        pointer-events: none;
        opacity: 0.5;
      }
    }
  }
}
