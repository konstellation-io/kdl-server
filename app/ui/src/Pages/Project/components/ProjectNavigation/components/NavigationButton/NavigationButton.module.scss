@import 'Styles/variables';
@import 'Styles/mixins';
@import 'Styles/colors';

.navButton {
  height: 5 * $grid-unit;
  display: flex;
  align-items: center;
  cursor: pointer;
  background-color: transparent;
  border-left: $grid-unit/4 solid transparent;

  transition: background-color ease 0.3s, border-left ease 0.3s, color ease 0.2s,
    opacity ease 0.3s;
  color: font-color(lowlight);

  .icon {
    margin: 0 1.4 * $grid-unit;
    transition: transform ease 0.3s;
  }

  .label {
    @include font-button;
    margin-left: $grid-unit/4;
    white-space: nowrap;
  }

  &:hover {
    background-color: palette(base, 400);
    color: font-color(light);
  }
}

.active .navButton {
  color: font-color(highlight);
  border-left: $grid-unit/4 solid palette(highlight);
}

.disabled {
  pointer-events: none;

  .navButton,
  .label,
  .icon {
    color: palette(lowlight, 300);
    opacity: 0.5;
  }
}

.collapsed .icon {
  transform: scaleX(-1);
}
