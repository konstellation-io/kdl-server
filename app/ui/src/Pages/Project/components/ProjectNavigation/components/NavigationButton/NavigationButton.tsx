import * as React from 'react';
import cx from 'classnames';
import styles from './NavigationButton.module.scss';
import { OverridableComponent } from '@material-ui/core/OverridableComponent';
import { SvgIconTypeMap } from '@material-ui/core';

export enum IconSize {
  SMALL = 'icon-small',
  REGULAR = 'icon-regular',
}

type Props = {
  label: string;
  Icon: OverridableComponent<SvgIconTypeMap>;
  title?: string;
  iconSize?: IconSize;
  onClick?: () => void;
  className?: string;
  disabled?: boolean;
  dataTestId?: string;
};

function NavigationButton({
  label,
  title,
  Icon,
  onClick,
  className = '',
  dataTestId = undefined,
  iconSize = IconSize.REGULAR,
  disabled = false,
}: Props) {
  return (
    <div
      className={cx(styles.navButton, className, {
        [styles.disabled]: disabled,
      })}
      title={title || label}
      onClick={onClick}
      data-testid={dataTestId}
    >
      <Icon className={cx(iconSize, styles.icon)} />
      <span className={styles.label}>{label}</span>
    </div>
  );
}

export default NavigationButton;
