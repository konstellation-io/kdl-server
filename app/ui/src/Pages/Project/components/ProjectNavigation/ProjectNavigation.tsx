import { NavLink, useParams } from 'react-router-dom';
import React, { FC, useState } from 'react';
import usePanel, { PanelType } from 'Graphql/client/hooks/usePanel';
import useWorkspace, { CONFIG } from 'Hooks/useWorkspace';

import IconCollapse from '@material-ui/icons/KeyboardBackspace';
import IconSettings from '@material-ui/icons/Settings';
import NavElements from './components/NavElements/NavElements';
import NavigationButton from './components/NavigationButton/NavigationButton';
import { PANEL_ID } from 'Graphql/client/models/Panel';
import { RouteProjectParams } from 'Constants/routes';
import cx from 'classnames';
import navButtonStyles from './components/NavigationButton/NavigationButton.module.scss';
import styles from './ProjectNavigation.module.scss';
import { useReactiveVar } from '@apollo/client';
import { primaryPanel } from 'Graphql/client/cache';
import { SETTINGS_PANEL_OPTIONS } from '../../panelSettings';

export const NavButtonLink: FC<any> = ({ children, ...props }) => {
  return (
    <NavLink
      {...props}
      activeClassName={navButtonStyles.active}
      className={cx({ [navButtonStyles.disabled]: props.disabled })}
      exact
    >
      {children}
    </NavLink>
  );
};

function ProjectNavigation() {
  const { projectId } = useParams<RouteProjectParams>();
  const [{ navigationOpened }, saveConfiguration] = useWorkspace(projectId);
  const [opened, setOpened] = useState(navigationOpened);

  const panelData = useReactiveVar(primaryPanel);
  const { openPanel: openSettings, closePanel: closeSettings } = usePanel(
    PanelType.PRIMARY,
    SETTINGS_PANEL_OPTIONS
  );

  function onToggleOpened() {
    setOpened(!opened);
    saveConfiguration(CONFIG.NAVIGATION_OPENED, !opened);
  }

  function toggleSettingsPanel() {
    const shouldOpen = !panelData || panelData.id !== PANEL_ID.SETTINGS;

    if (shouldOpen) openSettings();
    else closeSettings();
  }

  return (
    <div
      className={cx(styles.container, { [styles.opened]: opened })}
      data-testid="navigationBar"
    >
      <div className={styles.top}>
        <NavElements isOpened={opened} />
      </div>
      <div className={styles.bottom}>
        <NavigationButton
          dataTestId="toggleSettings"
          onClick={toggleSettingsPanel}
          label="Settings"
          Icon={IconSettings}
        />
        <NavigationButton
          dataTestId="toggleBar"
          label="Collapse"
          title={opened ? 'Collapse' : 'Expand'}
          Icon={IconCollapse}
          onClick={onToggleOpened}
          className={cx({
            [navButtonStyles.collapsed]: !opened,
          })}
        />
      </div>
    </div>
  );
}

export default ProjectNavigation;
