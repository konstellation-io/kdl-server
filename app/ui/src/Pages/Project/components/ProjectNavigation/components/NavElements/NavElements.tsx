import AnimateHeight from 'react-animate-height';
import CircularProgress from '@material-ui/core/CircularProgress';
import { GetMe } from 'Graphql/queries/types/GetMe';
import { NavButtonLink } from '../../ProjectNavigation';
import NavigationButton from '../NavigationButton/NavigationButton';
import PauseIcon from '@material-ui/icons/Pause';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import React from 'react';
import { RouteProjectParams } from 'Constants/routes';
import cx from 'classnames';
import styles from './NavElements.module.scss';
import { useParams } from 'react-router-dom';
import useProjectNavigation from 'Hooks/useProjectNavigation';
import { useQuery } from '@apollo/client';
import useTool from 'Graphql/hooks/useTool';
import ConfirmAction from 'Components/Layout/ConfirmAction/ConfirmAction';

import GetMeQuery from 'Graphql/queries/getMe';

type Props = {
  isOpened: boolean;
};

function NavElements({ isOpened }: Props) {
  const { projectId } = useParams<RouteProjectParams>();
  const { mainRoutes, userToolsRoutes, projectToolsRoutes } =
    useProjectNavigation(projectId);
  const { updateProjectActiveTools, projectActiveTools } = useTool();
  const { data, loading } = useQuery<GetMe>(GetMeQuery);
  const areToolsActive = data?.me.areToolsActive;

  if (loading) return null;

  function toggleTools() {
    updateProjectActiveTools(!areToolsActive);
  }

  function renderToggleToolsIcon() {
    if (projectActiveTools.loading)
      return () => (
        <CircularProgress className={styles.loadingTools} size={16} />
      );
    return areToolsActive ? PauseIcon : PlayArrowIcon;
  }

  return (
    <>
      {mainRoutes.map(({ Icon, label, route }) => (
        <NavButtonLink to={route} key={label}>
          <NavigationButton label={label} Icon={Icon} />
        </NavButtonLink>
      ))}
      <div className={styles.projectTools}>
        {projectToolsRoutes.map(({ Icon, label, route, disabled }) => (
          <NavButtonLink to={route} key={label} disabled={disabled}>
            <NavigationButton label={label} Icon={Icon} />
          </NavButtonLink>
        ))}
        <div
          className={cx(styles.userTools, {
            [styles.started]: areToolsActive,
            [styles.stopped]: !areToolsActive,
          })}
        >
          <AnimateHeight height={isOpened ? 'auto' : 0} duration={300}>
            <div className={styles.userToolLabel}>USER TOOLS</div>
          </AnimateHeight>
          {userToolsRoutes.map(({ Icon, label, route, disabled }) => (
            <NavButtonLink to={route} key={label} disabled={disabled}>
              <NavigationButton label={label} Icon={Icon} />
            </NavButtonLink>
          ))}
          <div
            className={cx(styles.toggleToolsWrapper, {
              [styles.disabled]: projectActiveTools.loading,
            })}
            data-testid="confirmationModal"
          >
            <ConfirmAction
              title="STOP YOUR TOOLS"
              subtitle="You are going to stop your user tools, please confirm your choice."
              action={toggleTools}
              actionLabel="STOP"
              skipConfirmation={!areToolsActive}
              warning
            >
              <NavigationButton
                dataTestId={areToolsActive ? 'stopTools' : 'runTools'}
                label={areToolsActive ? 'Stop tools' : 'Run tools'}
                Icon={renderToggleToolsIcon()}
              />
            </ConfirmAction>
          </div>
        </div>
      </div>
    </>
  );
}

export default NavElements;
