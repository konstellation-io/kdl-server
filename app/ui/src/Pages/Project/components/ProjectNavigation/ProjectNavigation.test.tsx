import * as React from 'react';
import ProjectNavigation, { NavButtonLink } from './ProjectNavigation';
import { shallow } from 'enzyme';
import { StaticRouter } from 'react-router';
import NavElements from './components/NavElements/NavElements';
import NavigationButton from './components/NavigationButton/NavigationButton';

jest.mock('react-router-dom', () => ({
  useRouteMatch: () => ({
    isExact: false,
  }),
  useParams: () => ({
    projectId: 'projectId',
  }),
}));

describe('ProjectNavigation component', () => {
  const component = shallow(<ProjectNavigation />);

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should be opened by default', () => {
    expect(component.find(NavElements).props().isOpened).toBeTruthy();
  });

  it('should be closed after clicking on the toggle button', () => {
    // Act.
    const toggleButton = component.find(NavigationButton).last();
    toggleButton.simulate('click');

    // Assert.
    expect(component.find(NavElements).props().isOpened).toBeFalsy();
  });
});

describe('NavButtonLink component', () => {
  const component = shallow(
    <StaticRouter>
      <NavButtonLink to="/home" disabled={false} />
    </StaticRouter>,
  );

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
