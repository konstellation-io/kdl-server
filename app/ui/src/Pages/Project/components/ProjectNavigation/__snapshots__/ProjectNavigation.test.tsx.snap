// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavButtonLink component Component match snapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  staticContext={Object {}}
>
  <NavButtonLink
    disabled={false}
    to="/home"
  />
</Router>
`;

exports[`ProjectNavigation component Component match snapshot 1`] = `
<div
  className="container opened"
  data-testid="navigationBar"
>
  <div
    className="top"
  >
    <NavElements
      isOpened={true}
    />
  </div>
  <div
    className="bottom"
  >
    <NavigationButton
      Icon={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      dataTestId="toggleSettings"
      label="Settings"
      onClick={[Function]}
    />
    <NavigationButton
      Icon={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      className=""
      dataTestId="toggleBar"
      label="Collapse"
      onClick={[Function]}
      title="Collapse"
    />
  </div>
</div>
`;
